### PR TITLE
nautilus: rbd: journal: fix flush by age and in-flight byte tracking

### PR DESCRIPTION
**Pull request by trociny**
backport tracker: https://tracker.ceph.com/issues/42953
backport tracker: https://tracker.ceph.com/issues/47887

---

backport of https://github.com/ceph/ceph/pull/31392
parent tracker: https://tracker.ceph.com/issues/42598

backport of https://github.com/ceph/ceph/pull/37699
parent tracker: https://tracker.ceph.com/issues/47880

this backport was staged using ceph-backport.sh version 16.0.0.6848
find the latest version at https://github.com/ceph/ceph/blob/master/src/script/ceph-backport.sh

### CODE REVIEW

File: src/journal/JournalRecorder.h [hunk 1]
```diff
@@ -38,6 +38,7 @@ class JournalRecorder {
 
 private:
   typedef std::map<uint8_t, ObjectRecorderPtr> ObjectRecorderPtrs;
+  typedef std::vector<std::unique_ptr<Mutex::Locker>> Lockers;
```
Comment (idryomov) [inline src/journal/JournalRecorder.h:41]: This ends up being `unique_ptr<lock_guard<Mutex>>`, correct?  Could `unique_lock<Mutex>` be used instead?

**Comment by idryomov** [comment]
This doesn't appear to be sufficient.  That assert is still being hit, as well as the new one introduced in this PR.  I re-opened https://tracker.ceph.com/issues/51100.

**Comment by trociny** [comment]
This is not backported yet and looks related (at least for one of the crashes) 458ab997fe77 https://tracker.ceph.com/issues/47880

File: src/journal/JournalRecorder.h (hunk 1 above)
Comment (trociny) [reply src/journal/JournalRecorder.h:41]: I was not able to make it compile. And I did not find examples of `unique_lock<Mutex>`  usage in that branch.

**Comment by trociny** [comment]
Backported 458ab99 here.

**Comment by idryomov** [comment]
The new crash is (expectedly) gone, but the original one persists:

http://qa-proxy.ceph.com/teuthology/dis-2021-06-09_14:08:54-rbd-wip-dis-testing-nautilus-distro-basic-smithi/6162474/teuthology.log

**Comment by tchaikov** [comment]
@trociny @idryomov JFYI, nautilus is locked.

**Comment by smithfarm** [comment]
@trociny What is the status of this PR? Should it be marked DNM?

**Comment by trociny** [comment]
The PR is probably still fine to merge. But I marked it DNM for now.

**Pull request closed without merging**